modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,171 @@
+// proportion of nodes or cores in the zone.
+// 2. This is compared with the proportion of endpoints in the zone to calculate
+// the deviation from the expected number of endpoints in a perfectly balanced
+// system.
+// 3. EndpointSlices in zones with less endpoints than expected will receive
+// endpoints from zones that have more endpoints than expected.
+type LocalSliceAlgorithm struct{}
+
+type endpointDeviation struct {
+	name      string
+	deviation int
+}
+
+// endpointsList maintains the order of the endpoints available/needed, used as
+// a list
+type endpointsList struct {
+	byZone []endpointDeviation
+}
+
+// push to back
+func (el *endpointsList) push(newValue endpointDeviation) {
+	el.byZone = append(el.byZone, newValue)
+}
+
+// pop the front
+func (el *endpointsList) pop() {
+	if len(el.byZone) == 0 {
+		return
+	}
+	el.byZone = el.byZone[1:]
+}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy
+func (alg LocalSliceAlgorithm) CreateSliceGroups(region regionInfo) (map[string]EndpointSliceGroup, error) {
+	if region.zoneDetails == nil {
+		return nil, fmt.Errorf("zoneDetail should not be nil")
+	}
+	sliceGroups := map[string]EndpointSliceGroup{}
+	// endpointsAvailable stores 'endpointDeviation' -- zones with 'extra'
+	// endpoints
+	endpointsAvailable := endpointsList{}
+	// endpointsNeeded stores zones with number of endpoints needed
+	endpointsNeeded := endpointsList{}
+	for zoneName, zone := range region.zoneDetails {
+		var localGroup EndpointSliceGroup
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.Weights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]int{}
+
+		// the expected number of endpoints is with the same ratio as the number
+		// of nodes in this zone
+		expectedEndpoints := int(zone.nodesRatio * float64(region.totalEndpoints))
+		// deviation: - means need more endpoints from other sliceGroups, +
+		// means need give out endpoints to other sliceGroups
+		deviation := zone.Endpoints - expectedEndpoints
+		if deviation > 0 {
+			endpointsAvailable.push(endpointDeviation{name: zoneName, deviation: deviation})
+			localGroup.Composition[zoneName] = int(expectedEndpoints)
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -deviation})
+			localGroup.Composition[zoneName] = zone.Endpoints
+		}
+
+		sliceGroups[zoneName] = localGroup
+	}
+
+	err := balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, sliceGroups)
+	if err != nil {
+		return nil, err
+	}
+	return sliceGroups, nil
+}
+
+// balanceSliceGroups helps to 'rent' endpoints from zones have extra endpoints
+// to EndpointSliceGroups that have insufficient ones
```

Comment:
Maybe `balanceSliceGroups distributes endpoints from zones with extra endpoints to EndpointSliceGroups for zones with insufficient endpoints`.

Reply:
done